cirq/google/known_devices.py

```diff
@@ -47,6 +54,64 @@ def _parse_device(s: str) -> Tuple[List[GridQubit], Dict[str, Set[GridQubit]]]:
     return qubits, measurement_lines
 
 
+def create_device_proto_from_diagram(
+        s: str,
+        gate_set: serializable_gate_set.SerializableGateSet = None,
+        durations: Dict[str, int] = None,
+) -> device_pb2.DeviceSpecification:
+    """
+    Parse ASCIIart device layout into DeviceSpecification proto containing
+    information about qubits and targets.  Note that this does not populate
+    the gate set information of the proto.
+
+    Also accepts a gate set object that will create the gate set proto.
+
+    This function also assumes that only adjacent qubits can be targets.
+    """
+    qubits, _ = _parse_device(s)
+    spec = device_pb2.DeviceSpecification()
+    meas_targets = spec.valid_targets.add()
+    meas_targets.name = _MEAS_TARGET_SET
+    meas_targets.target_ordering = device_pb2.TargetSet.SUBSET_PERMUTATION
+
+    grid_targets = spec.valid_targets.add()
+    grid_targets.name = _2_QUBIT_TARGET_SET
+    grid_targets.target_ordering = device_pb2.TargetSet.SYMMETRIC
+    qubit_set = frozenset(qubits)
+    neighbor_set: Set[Tuple] = set()
+    spec.valid_qubits.extend([q.proto_id() for q in qubits])
+    for q in qubits:
+        for neighbor in q.neighbors(qubit_set):
+            if tuple((neighbor, q)) not in neighbor_set:
+                # Don't add pairs twice
+                new_target = grid_targets.targets.add()
+                new_target.ids.extend([q.proto_id(), neighbor.proto_id()])
+                neighbor_set.add(tuple((q, neighbor)))
+
+    if gate_set is not None:
+        gs_proto = spec.valid_gate_sets.add()
+        gs_proto.name = gate_set.gate_set_name
+        gate_ids: Set[str] = set()
+        for gate_type in gate_set.serializers:
+            for serializer in gate_set.serializers[gate_type]:
+                gate_id = serializer.serialized_gate_id
+                if gate_id in gate_ids:
+                    # Only add each type once
+                    continue
+
+                gate_ids.add(gate_id)
+                gate = gs_proto.valid_gates.add()
+                gate.id = gate_id
+                if gate_type == MeasurementGate:
+                    gate.valid_targets.extend([_MEAS_TARGET_SET])
+                else:
+                    gate.valid_targets.extend([_2_QUBIT_TARGET_SET])
+                if durations is not None and gate.id in durations:
+                    gate.gate_duration_picos = durations[gate.id]
+                # TODO: add valid args and number of qubits
```

Comment:
I wonder if it would be better to just generate `.pbtxt` versions of the descriptions for existing known devices and check those it, rather than try to generate these from grids where we don't have all the needed information (like gate arity and args).

Reply:
Yeah, that's my plan.  ... However, how are we going to generate the .pbtxt versions?

The goal is that this (or something similar to it, will help to generate the .pbtxt version.  I certainly don't want to do it by hand.

I think, with the serializers and the gate set, we have the necessary information (for the use cases we have anyway).w